crates/lang/src/parse.rs

```diff
@@ -549,13 +549,180 @@ fn seek_to<'a, R: pest::RuleType>(
     None
 }
 
+#[derive(Debug, PartialEq)]
+pub struct RemoteDbInfo {
+    addrs: Vec<RemoteAddr>,
+    username: Option<String>,
+    password: Option<String>,
+    database_name: String,
+    table_name: String,
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct RemoteAddr {
+    ip_addr: Option<IpAddr>,
+    host_name: Option<String>,
+    port: Option<u16>,
+}
+
+impl Default for RemoteDbInfo {
+    fn default() -> Self {
+        Self {
+            addrs: vec![],
+            username: None,
+            password: None,
+            database_name: "default".to_string(),
+            table_name: "".to_string(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TablePlaceKind {
+    Local,
+    Remote(RemoteDbInfo),
+}
+
+#[derive(Debug)]
+pub struct TablePlaceKindContext {
+    pub place_kind: TablePlaceKind,
+}
+
+impl TablePlaceKindContext {
+    fn mut_remote_info<'a>(&'a mut self) -> Option<&'a mut RemoteDbInfo> {
+        if let TablePlaceKind::Remote(place_kind) = &mut self.place_kind {
+            Some(place_kind)
+        } else {
+            None
+        }
+    }
+}
+
+pub fn parse_table_place(pair: Pair<Rule>) -> LangResult<TablePlaceKindContext> {
+    let mut ctx = TablePlaceKindContext {
+        place_kind: TablePlaceKind::Local,
+    };
+    ctx.parse(pair)?;
+    Ok(ctx)
+}
+
+#[inline]
+fn parse_ip(pair: Pair<Rule>) -> LangResult<Option<IpAddr>> {
+    for p in pair.into_inner() {
+        match p.as_rule() {
+            Rule::ipv4_lit | Rule::ipv6_lit => {
+                let ip: IpAddr = p
+                    .as_str()
+                    .parse()
+                    .map_err(|e| LangError::WrappingParseIpAddrError(e))?;
+                return Ok(Some(ip));
+            }
+            _ => {}
+        }
+    }
+    Ok(None)
+}
+
+#[inline]
+fn parse_ip_address(pair: Pair<Rule>) -> LangResult<RemoteAddr> {
+    let mut addr = RemoteAddr::default();
+
+    for p in pair.into_inner() {
+        match p.as_rule() {
+            Rule::ip => {
+                addr.ip_addr = parse_ip(p)?;
+            }
+            Rule::port => {
+                let port: u16 = p
+                    .as_str()
+                    .parse()
+                    .map_err(|e| LangError::WrappingParseIntError(e))?;
+                addr.port = Some(port)
+            }
+            _ => {}
+        }
+    }
+
+    Ok(addr)
+}
+
+#[inline]
+fn parse_host_address(pair: Pair<Rule>) -> RemoteAddr {
+    let s = pair.as_str().trim().to_string();
+    let vs: Vec<&str> = s.split(":").collect();
+    let mut addr = RemoteAddr::default();
+
+    addr.host_name = Some(vs[0].to_owned());
+    if vs.len() == 2 {
```

Comment:
if len >2, do we return an error?

Reply:
The port can be default, when it's default port, it will be None, we can handle default port in the runtime.

Reply:
reasonable!